### PR TITLE
Fix API leaking Usermail if not logged in

### PR DESCRIPTION
**Pull request by JakobDev**
The API should only return the real Mail of a User, if the caller is logged in. The check do to this don't work. This PR fixes this. This not really a security issue, but can lead to Spam.

### CODE REVIEW

File: services/convert/user.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ func toUser(ctx context.Context, user *user_model.User, signed, authed bool) *ap
 		ID:          user.ID,
 		UserName:    user.Name,
 		FullName:    user.FullName,
-		Email:       user.GetEmail(),
+		Email:       user.GetPlaceholderEmail(),
```
Comment (lunny) [inline services/convert/user.go:54]: So we need to check if signed is true?

Comment (JakobDev) [reply services/convert/user.go:54]: The file already cntians this code down below which does the check:
```go
// hide primary email if API caller is anonymous or user keep email private
if signed && (!user.KeepEmailPrivate || authed) {
	result.Email = user.Email
}
```
As `user.GetMail()` only checks if the Mail is private, but not if someone is logged in, it always returns the Mail if it is not private, so this Check hasn't worked. Now that `Email` is set to the Placeholder before this Check, it is now working.

**Comment by jpraet** [comment]
I would argue that this _is_ in fact a security issue?

On try.gitea.io I cannot see user emails on the explore or user profile pages, likely because ui.SHOW_USER_EMAIL setting is set to false?

But a non-logged in user can retrieve the emails via the API e.g. https://try.gitea.io/api/v1/users/lunny.

On top of checking whether a user is logged in, I think the API should probably also respect the ui.SHOW_USER_EMAIL setting (although.. it's a "ui" setting, so I'm not sure)?

And the is_admin property should maybe also not be included in the API response if the caller is not admin themselves, as to not advertise to attackers which users are admin?

**Review by silverwind — APPROVED** [review]

**Review by puni9869 — APPROVED** [review]
🚀

**Comment by silverwind** [comment]
Test failures related.

**Comment by JakobDev** [comment]
> Test failures related.

It looks like this test use the Web. It doesn't use any API, but only the API is touched by this PR. Interesting. I need to take a look at this.

**Comment by silverwind** [comment]
In some rare cases web does use the API, so this might be one of those cases.

**Comment by JakobDev** [comment]
`TestXSSUserFullName` passes with SQLite on my local PC

**Comment by JakobDev** [comment]
I was stupid. ` TestXSSUserFullName` was the last test shown in the CI, so I though this test was broken, but it was another who is actually using the API.

**Comment by JakobDev** [comment]
This is a Bugfix, so it should probably backported

**Comment by wxiaoguang** [comment]
Added backport tag

**Comment by vorlon** [comment]
I agree with jpraet's [comment](https://github.com/go-gitea/gitea/pull/25097#issuecomment-1579149601) above and would definitely consider this a security issue, as exposing the e-mail address via API when it is not visible via the UI is something that is probably not expected by a user. 
As a security issue and considering gitea is a CNA for CVE IDs, a CVE ID should probably be issued and included on https://about.gitea.com/security/

**Pull request merged**